src/MatrixProductOperator.jl

```diff
@@ -0,0 +1,171 @@
+#using OptimizedEinsum: ContractionPath
+using UUIDs: uuid4
+using IterTools: partition
+using Random
+
+abstract type MatrixProductOperator{B} <: Operator{B} end
+
+function MatrixProductOperator(arrays; bounds::Type{<:Bounds} = Open, kwargs...)
+    MatrixProductOperator{bounds}(arrays; kwargs...)
+end
+
+function MatrixProductOperator{Open}(arrays; χ = nothing, order = (:l, :r, :i, :o), meta...)
+    !issetequal(order, (:l, :r, :i, :o)) && throw(ArgumentError("`order` must be a permutation of the :l, :r, :i and :o"))
+    order = Dict(dim => i for (i, dim) in enumerate(order))
+
+    # check format
+    # NOTE matching index dimensions are checked on `push!(::TensorNetwork,...)`
+    # TODO check χ?
+    if !all(==(4) ∘ ndims, arrays[2:end-1]) || ndims(first(arrays)) != 3 || ndims(last(arrays)) != 3
+        throw(DimensionMismatch("virtual bond-dims mismatch"))
+    end
+
+    tn = TensorNetwork{MatrixProductOperator{Open}}(; χ = χ, order = order, meta...)
+
+    n = length(arrays)
+    vinds = Dict(x => Symbol(uuid4()) for x in partition(1:n, 2, 1))
+    oinds = Dict(i => Symbol(uuid4()) for i in 1:n)
+    iinds = Dict(i => Symbol(uuid4()) for i in 1:n)
+    permutator = [order[i] for i in (:l, :r, :i, :o)]
+    # add boundary tensors
+    # first
+    labels = permute!([vinds[(1,2)], iinds[1], oinds[1]],  normalizeperm!([order[:r], order[:i], order[:o]]))
+    push!(tn, Tensor(first(arrays), labels))
+
+    # last
+    labels = permute!([vinds[(n-1,n)], iinds[n], oinds[n]], normalizeperm!([order[:l], order[:i], order[:o]]))
+    push!(tn, Tensor(last(arrays), labels))
+
+    # add other tensors
+    for (i, data) in zip(2:n-1, arrays[2:end-1])
+        lind = vinds[(i - 1, i)]
+        rind = vinds[(i, i + 1)]
+
+        labels = [lind, rind, iinds[i], oinds[i]]
+        permute!(labels, permutator)
```

Comment:
Same problem here. We were lucky in the tests because in this case we considered the permutation was equal to the inverse permutation. Now this will work in general.
```suggestion
        invpermute!(labels, permutator)
```